backend/src/users.rs

```diff
@@ -0,0 +1,179 @@
+pub struct CreateUserSchema {
+    pub username: String,
+    pub first_name: String,
+    pub last_name: String,
+    pub email: String,
+    pub password: String,
+}
+
+// Likely want to add 'Optional' fields for last name
+// If Optional fields added, change .fetch_* to .fetch_optional(...)
+#[derive(Debug, Clone, Deserialize, Serialize, FromRow, FromForm)]
+pub struct UserModel {
+    pub id: i64,
+    pub username: String,
+    pub first_name: String,
+    pub last_name: String,
+    pub email: String,
+    pub password: String,
+}
+
+#[derive(Debug)]
+pub enum SignupError {
+    NonUniqueIdError,
+    UnknownQueryError,
+    UnknownDatabaseError,
+}
+
+impl From<sqlx::Error> for SignupError {
+    fn from(error: sqlx::Error) -> Self {
+        match error {
+            sqlx::Error::Database(db_error) => {
+                let pg_error = db_error.downcast::<sqlx::postgres::PgDatabaseError>();
+                match pg_error.code() {
+                    "23505" => {
+                        println!("Duplicate user ID.");
+                        SignupError::NonUniqueIdError
+                    }
+                    _ => {
+                        println!("-- An error the server didn't account for --");
+                        println!("{:?}", pg_error);
+                        SignupError::UnknownQueryError
+                    }
+                }
+            }
+            _ => {
+                println!("Something else happened");
+                SignupError::UnknownDatabaseError
+            }
+        }
+    }
+}
+
+impl std::fmt::Display for SignupError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            SignupError::NonUniqueIdError => {
+                write!(f, "Duplicate username or email contained a duplicate key.")
+            }
+            SignupError::UnknownQueryError => {
+                write!(f, "Database query contained an unspecified error.")
+            }
+            SignupError::UnknownDatabaseError => write!(f, "Database error, not query related."),
+        }
+    }
+}
+
+pub struct CORS;
+#[rocket::async_trait]
+impl fairing::Fairing for CORS {
+    fn info(&self) -> fairing::Info {
+        fairing::Info {
+            name: "Add CORS headers to responses",
+            kind: fairing::Kind::Response,
+        }
+    }
+
+    async fn on_response<'r>(&self, request: &'r Request<'_>, response: &mut Response<'r>) {
+        response.set_header(http::Header::new(
+            "Access-Control-Allow-Origin",
+            request.headers().get_one("Origin").unwrap_or("*"),
+        ));
+        response.set_header(http::Header::new(
+            "Access-Control-Allow-Methods",
+            "POST, GET, PATCH, OPTIONS",
+        ));
+        response.set_header(http::Header::new("Access-Control-Allow-Headers", "*"));
+        response.set_header(http::Header::new(
+            "Access-Control-Allow-Credentials",
+            "true",
+        ));
+    }
+}
```

Comment:
This could/should be in the main file or in its own module, CORS is an app level thing

Reply:
4aa22fd07d12781bb6801b7898d735f648de7193